uaclient/apt.py

```diff
@@ -104,6 +104,48 @@ def run_apt_command(
     except util.ProcessExecutionError as e:
         if "Could not get lock /var/lib/dpkg/lock" in str(e.stderr):
             error_msg += " Another process is running APT."
+
+        """
+        Treat errors where one of the APT repositories
+        is invalid or unreachable. In that situation, we alert
+        which repository is causing the error
+        """
+        rep_error = "The repository"
+        fetch_error = "Failed to fetch"
+        if rep_error in str(e.stderr) or fetch_error in str(e.stderr):
```

Comment:
1. this bit of logic should be a separate testable function with a couple of simple unit tests.
2. We should probably just walk the lines of the stderr anyway and use re.match on it:
3. let's use named matching regex pattern so we can easily grab the matched object by keyname
```
failed_repos = set()
for line in e.stderr.splitlines():
   pattern_match = re.search(r"(Failed to fetch|The repository) (?P<url>[^\s]+)(?= )", line)
   if pattern_match:
       # Xenial emits both 'Failed to fetch' and 'The repository' errors. Set avoids url dupes.
       failed_urls.update(["- " + pattern_match.groupdict("url")])
   if failed_urls:
       error_msg += status.ERROR_APT_CONFIG_URLS.format(urls="\n".join(failed_urls))

Reply:
@blackboxsw That's a good point to move that logic into a separate function, however the set directive will not work here because, for Xenial, the urls are not exactly the same. For example, we will have those two lines in xenial:

```
The repository 'http://ppa.launchpad.net/ua-client/proposed/ubun 404' ...
Failed to fetch http://ppa.launchpad.net/ua-client/proposed/ubun/dists/xenial/main/binary-amd64/Packages
```

So the URLs are not exactly the same. That's why I am using the for loop here.

But I will move the logic into a separate function